tools/python/upload_and_run_browserstack_tests.py

```diff
@@ -0,0 +1,121 @@
+import argparse
+import os
+import time
+
+import requests
+
+
+def response_to_json(response):
+    response.raise_for_status()
+    response_json = response.json()
+    print(response_json)
+    print("-" * 30)
+    return response_json
+
+
+def upload_apk_parse_json(post_url, apk_path, id, token):
+    with open(apk_path, "rb") as apk_file:
+        response = requests.post(post_url, files={"file": apk_file}, auth=(id, token))
+    return response_to_json(response)
+
+
+def browserstack_build_request(devices, app_url, test_suite_url, test_platform, id, token):
+    headers = {}
+
+    json_data = {
+        "devices": devices,
+        "app": app_url,
+        "testSuite": test_suite_url,
+    }
+
+    build_response = requests.post(
+        f"https://api-cloud.browserstack.com/app-automate/{test_platform}/v2/build",
+        headers=headers,
+        json=json_data,
+        auth=(id, token),
+    )
+
+    return response_to_json(build_response)
+
+
+def build_query_loop(build_id, test_platform, id, token):
+    """Called after a build is initiated on Browserstack. Repeatedly queries the REST endpoint to check for the
+    status of the tests in 30 second intervals.
+    """
+    tests_status = "running"
+
+    while tests_status == "running":
+        time.sleep(30)
+
+        test_response = requests.get(
+            f"https://api-cloud.browserstack.com/app-automate/{test_platform}/v2/builds/{build_id}",
+            auth=(id, token),
+        )
```

Comment:
I added an overall timeout of 15 minutes to the script step in the yaml file. Let me know if this should we swapped to timeouts for individual steps?

Reply:
Two potential use cases though - in the CI and directly by a developer. Adding one to the `get` means that won't be allowed to hang for an unreasonable amount of time in both.

We should still keep the yaml one for the overall timeout in a CI though as that covers everything the script does. It could get stuck on 'running' for a long time with the `get` being short each time.